notebooks/03-query/11-joins.ipynb

```diff
@@ -0,0 +1,2475 @@
+      "2023-10-26 01:00:04.285867 <<<<<<<<<<<<< \n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "datetime.datetime(2023, 10, 26, 1, 0, 4, 285867)"
+      ]
+     },
+     "execution_count": 11,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# define user_age_group lookup by building it from a query of users\n",
+    "sql='''\n",
+    "SELECT user_id, ROUND( TIMESTAMPDIFF(YEAR, TIME_PARSE(\"dob\"), CURRENT_TIMESTAMP),-1) as age_group\n",
+    "FROM \"users\"\n",
+    "'''\n",
+    "results = sql_client.sql(sql)\n",
+    "map = dict( zip([ r['user_id'] for r in results ], [ r['age_group'] for r in results ]))\n",
+    "\n",
+    "lookup_tier = \"__default\"\n",
+    "lookup_name = \"user_age_group\"\n",
+    "lookup_definition_version = datetime.now().strftime(\"%Y-%m-%dT%H:%M:%SZ\")\n",
+    "lookup_definition = {\n",
+    "    lookup_tier: {\n",
+    "        lookup_name: {\n",
+    "            \"version\": lookup_definition_version,  \n",
+    "            \"lookupExtractorFactory\": {\n",
+    "                \"type\": \"map\",\n",
+    "                    \"map\": map\n",
+    "            }\n",
+    "        }\n",
+    "    }\n",
+    "}\n",
+    "postLookup(lookup_definition)\n",
+    "waitForLookup(lookup_tier, lookup_name, 30)\n",
+    "print_timestamp()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "6bf7f21b-7cdd-460e-bc00-e74037e21717",
+   "metadata": {},
+   "source": [
+    "## Prep Query Testing and Measurement Function\n",
+    "We measure each method of joining the data by running the query multiple times and avoiding cacheing of the results in order to get a avg, min and max runtimes of each query. We'll use the following function to achieve this:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 12,
+   "id": "38f8b68a-52ca-41bf-9d84-ea668632528c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from statistics import mean \n",
+    "\n",
+    "def measure_query( sql: str, iterations: int ):\n",
+    "    req = sql_client.sql_request(sql)\n",
+    "    req.add_context(\"populateCache\", \"false\")  # run without cacheing results to get a real sense of performance\n",
+    "    req.add_context(\"useCache\", \"false\")  # do not use cached results\n",
+    "    stats = []\n",
+    "    while (iterations>0):\n",
+    "      start = datetime.now()\n",
+    "      sql_client.sql(req)\n",
+    "      end = datetime.now()\n",
+    "      stats.append( (end - start).total_seconds() * 10**3 ) # add run time in milliseconds\n",
+    "      iterations -=1\n",
+    "    return f\"Results = avg:{mean(stats)} ms   min:{min(stats)} ms  max:{max(stats)} ms\""
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "id": "4cf7592d-b852-4712-b8ec-60be9df1224d",
+   "metadata": {},
+   "source": [
+    "## Synchronous Queries on Scatter/Gather Engine\n",
```

Comment:
All of this is documentation - @techdocsmith @vtlim how could we incorporate this there?

Reply:
This is more like a User Guide which does provide some explanation. I'm trying to provide summaries of useful concepts here so that I can expand on them. While the idea of multiple scatter/gather steps is described in the docs here: https://druid.apache.org/docs/latest/querying/query-execution#join. I think a deeper explanation is needed to understand the effects of that in different scenarios.